File: python/cdp/x402/core/balance_check.py
"""Client-side pre-flight balance check for x402 payments."""

from __future__ import annotations

from collections.abc import Callable, Coroutine
from typing import TYPE_CHECKING, Any

from web3 import AsyncWeb3
from web3.providers import AsyncHTTPProvider

from cdp.x402.core.constants import (
    BASE_MAINNET_CAIP2,
    BASE_SEPOLIA_CAIP2,
    CDP_EVM_RPC_URLS,
    SOLANA_DEVNET_CAIP2,
    SOLANA_MAINNET_CAIP2,
)
from cdp.x402.core.guardrails.normalize import normalize_asset
from cdp.x402.core.utils import default_warn, read_required_atomic

if TYPE_CHECKING:
    from cdp import CdpClient

_CAIP_TO_CDP_EVM_NETWORK: dict[str, str] = {
    BASE_MAINNET_CAIP2: "base",
    BASE_SEPOLIA_CAIP2: "base-sepolia",
}

_CAIP_TO_CDP_SVM_NETWORK: dict[str, str] = {
    SOLANA_MAINNET_CAIP2: "solana",
    SOLANA_DEVNET_CAIP2: "solana-devnet",
}


class InsufficientFundsError(Exception):
    """Raised when the wallet's balance is below the required payment amount."""

    code: str = "insufficient_funds"

    def __init__(
        self,
        *,
        required: int,
        available: int,
        asset: str,
        network: str,
        address: str,
    ) -> None:
        super().__init__(
            f"Insufficient funds for x402 payment: wallet {address} on {network} has "
            f"{available} of asset {asset}, but {required} is required."
        )
        self.required = required
        self.available = available
        self.asset = asset
        self.network = network
        self.address = address


async def _get_evm_balance(
    cdp_client: Any,
    address: str,
    network: str,
    asset: str,
) -> int:
    page_token: str | None = None
    while True:
        page = await cdp_client.evm.list_token_balances(
            address=address,
            network=network,
            page_token=page_token,
        )
        for entry in page.balances:
            if normalize_asset(entry.token.contract_address) == asset:
                return int(entry.amount.amount)
        page_token = page.next_page_token
        if not page_token:
            break
    return 0


_ERC20_BALANCE_OF_ABI = [
    {
        "name": "balanceOf",
        "type": "function",
        "inputs": [{"name": "owner", "type": "address"}],
        "outputs": [{"name": "", "type": "uint256"}],
        "stateMutability": "view",
    }
]


async def _read_onchain_erc20_balance(
    rpc_url: str,
    token_address: str,
    wallet_address: str,
) -> int:
    w3 = AsyncWeb3(AsyncHTTPProvider(rpc_url))
    try:
        contract = w3.eth.contract(
            address=AsyncWeb3.to_checksum_address(token_address),
            abi=_ERC20_BALANCE_OF_ABI,
        )
        result: int = await contract.functions.balanceOf(
            AsyncWeb3.to_checksum_address(wallet_address)
        ).call()
        return result
    finally:
        await w3.provider.disconnect()


async def _get_svm_balance(
    cdp_client: Any,
    address: str,
    network: str,
    asset: str,
) -> int:
    page_token: str | None = None
    while True:
        page = await cdp_client.solana.list_token_balances(
            address=address,
            network=network,
            page_token=page_token,
        )
        for entry in page.balances:
            if entry.token.mint_address == asset:
                return int(entry.amount.amount)
        page_token = page.next_page_token
        if not page_token:
            break
    return 0


_WarnFn = Callable[[str, object | None], None]


def create_balance_check_hook(
    *,
    cdp_client: CdpClient,
    evm_address: str,
    svm_address: str,
    on_warning: _WarnFn | None = None,
    rpc_urls: dict[str, str] | None = None,
) -> Callable[[Any], Coroutine[Any, Any, None]]:
    """Build a ``BeforePaymentCreationHook`` that performs a pre-flight balance check."""
    warn = on_warning or default_warn
    resolved_rpc_urls: dict[str, str] = (
        {**CDP_EVM_RPC_URLS, **rpc_urls} if rpc_urls else CDP_EVM_RPC_URLS
    )

    async def hook(context: Any) -> None:
        req = context.selected_requirements
        required = read_required_atomic(req)
        if required is None or required == 0:
            return None

        network: str = req.network if hasattr(req, "network") else ""
        asset: str = req.asset if hasattr(req, "asset") else ""
        normalized_asset = normalize_asset(asset)

        evm_network = _CAIP_TO_CDP_EVM_NETWORK.get(network)
        evm_rpc_url = resolved_rpc_urls.get(network)
        svm_network = _CAIP_TO_CDP_SVM_NETWORK.get(network)

        available: int
        address: str

        asset_label = asset if svm_network else normalized_asset

        try:
            if evm_network:
                address = evm_address
                available = await _get_evm_balance(
                    cdp_client,
                    address=evm_address,
                    network=evm_network,
                    asset=normalized_asset,
                )
            elif evm_rpc_url:
                if not normalized_asset.startswith("0x"):
                    return None
                address = evm_address
                available = await _read_onchain_erc20_balance(
                    evm_rpc_url,
                    token_address=normalized_asset,
                    wallet_address=evm_address,
                )
            elif svm_network:
                address = svm_address
                available = await _get_svm_balance(
                    cdp_client,
                    address=svm_address,
                    network=svm_network,
                    asset=asset,
                )
            else:
                return None
        except Exception as exc:
            warn(
                f"Pre-flight balance check failed for {network} ({asset_label}); "
                "proceeding without check.",
                exc,
            )
            return None

        if available >= required:
            return None

        raise InsufficientFundsError(
            required=required,
            available=available,
            asset=asset,
            network=network,
            address=address,
        )

    return hook

File: typescript/packages/cdp-sdk/src/x402/balance-check.ts
/*
 * Client-side pre-flight balance check.
 *
 * Wires a BeforePaymentCreationHook that queries the CDP wallet's token
 * balance and fails fast with InsufficientFundsError when the balance is
 * below the required amount. The check is best-effort and silently skipped
 * for unsupported networks or on API errors.
 */
import { createPublicClient, http } from "viem";

import {
  CDP_EVM_RPC_URLS,
  baseMainnetCaip2,
  baseSepoliaCaip2,
  solanaDevnetCaip2,
  solanaMainnetCaip2,
} from "./constants.js";
import { tryParseAtomicFromRequirement } from "./guardrails/apply.js";
import { normalizeAsset } from "./guardrails/normalize.js";

import type { CdpClient } from "../client/cdp.js";
import type { BeforePaymentCreationHook } from "@x402/core/client";

const CAIP_TO_CDP_EVM_BALANCE_NETWORK: Record<string, "base" | "base-sepolia"> = {
  [baseMainnetCaip2]: "base",
  [baseSepoliaCaip2]: "base-sepolia",
};

const ERC20_BALANCE_OF_ABI = [
  {
    name: "balanceOf",
    type: "function",
    inputs: [{ name: "owner", type: "address" }],
    outputs: [{ name: "", type: "uint256" }],
    stateMutability: "view",
  },
] as const;

const CAIP_TO_CDP_SVM_BALANCE_NETWORK: Record<string, "solana" | "solana-devnet"> = {
  [solanaMainnetCaip2]: "solana",
  [solanaDevnetCaip2]: "solana-devnet",
};

/**
 * Thrown when the configured wallet's balance is below the amount required by
 * the server's `PaymentRequirements`.
 */
export class InsufficientFundsError extends Error {
  /** - Machine-readable error code. */
  readonly code = "insufficient_funds" as const;

  /** - Required amount in atomic units. */
  readonly required: bigint;

  /** - Available balance in atomic units. */
  readonly available: bigint;

  /** - Asset identifier. */
  readonly asset: string;

  /** - CAIP-2 network identifier. */
  readonly network: string;

  /** - Wallet address that was checked. */
  readonly address: string;

  /**
   * Creates an InsufficientFundsError with the given parameters.
   *
   * @param params - Error parameters.
   * @param params.required - Required amount in atomic units.
   * @param params.available - Available balance in atomic units.
   * @param params.asset - Asset identifier.
   * @param params.network - CAIP-2 network identifier.
   * @param params.address - Wallet address.
   */
  constructor(params: {
    required: bigint;
    available: bigint;
    asset: string;
    network: string;
    address: string;
  }) {
    super(
      `Insufficient funds for x402 payment: wallet ${params.address} on ${params.network} has ` +
        `${params.available} of asset ${params.asset}, but ${params.required} is required.`,
    );
    this.name = "InsufficientFundsError";
    this.required = params.required;
    this.available = params.available;
    this.asset = params.asset;
    this.network = params.network;
    this.address = params.address;
    Object.setPrototypeOf(this, InsufficientFundsError.prototype);
  }
}

interface BalanceCheckOptions {
  cdpClient: CdpClient;
  evmAddress: `0x${string}`;
  svmAddress: string;
  rpcUrls?: Partial<Record<string, { rpcUrl: string }>>;
  onWarning?: (message: string, cause?: unknown) => void;
}

const defaultWarning = (message: string, cause?: unknown): void => {
  if (cause === undefined)
    // eslint-disable-next-line no-console
    console.warn(`[@coinbase/cdp-sdk/x402] ${message}`);
  // eslint-disable-next-line no-console
  else console.warn(`[@coinbase/cdp-sdk/x402] ${message}`, cause);
};

const readOnChainErc20Balance = async (
  rpcUrl: string,
  contractAddress: `0x${string}`,
  walletAddress: `0x${string}`,
): Promise<bigint> => {
  const client = createPublicClient({ transport: http(rpcUrl) });
  return client.readContract({
    address: contractAddress,
    abi: ERC20_BALANCE_OF_ABI,
    functionName: "balanceOf",
    args: [walletAddress],
  });
};

const getEvmBalance = async (
  cdpClient: CdpClient,
  params: { address: `0x${string}`; network: "base" | "base-sepolia"; asset: string },
): Promise<bigint> => {
  let pageToken: string | undefined;
  do {
    const page = await cdpClient.evm.listTokenBalances({
      address: params.address,
      network: params.network,
      pageToken,
    });
    for (const entry of page.balances) {
      if (normalizeAsset(entry.token.contractAddress) === params.asset) {
        return entry.amount.amount;
      }
    }
    pageToken = page.nextPageToken;
  } while (pageToken);
  return 0n;
};

const getSvmBalance = async (
  cdpClient: CdpClient,
  params: { address: string; network: "solana" | "solana-devnet"; asset: string },
): Promise<bigint> => {
  let pageToken: string | undefined;
  do {
    const page = await cdpClient.solana.listTokenBalances({
      address: params.address,
      network: params.network,
      pageToken,
    });
    for (const entry of page.balances) {
      if (entry.token.mintAddress === params.asset) {
        return entry.amount.amount;
      }
    }
    pageToken = page.nextPageToken;
  } while (pageToken);
  return 0n;
};

/**
 * Builds a `BeforePaymentCreationHook` that performs a pre-flight balance
 * check against the configured CDP wallet.
 *
 * The hook is intentionally lenient: any condition under which the check
 * cannot be performed is treated as a pass so legitimate payments are never
 * blocked on pre-check noise.
 *
 * @param opts - Balance check configuration including the CDP client and wallet addresses.
 * @returns A BeforePaymentCreationHook that validates wallet balance before each payment.
 */
export function createBalanceCheckHook(opts: BalanceCheckOptions): BeforePaymentCreationHook {
  const { cdpClient, evmAddress, svmAddress } = opts;
  const warn = opts.onWarning ?? defaultWarning;
  const rpcUrls: Record<string, { rpcUrl: string }> = opts.rpcUrls
    ? ({ ...CDP_EVM_RPC_URLS, ...opts.rpcUrls } as Record<string, { rpcUrl: string }>)
    : CDP_EVM_RPC_URLS;

  return async context => {
    const req = context.selectedRequirements;
    const required = tryParseAtomicFromRequirement(req);
    if (required === undefined) return undefined;
    if (required === 0n) return undefined;

    const normalizedAsset = normalizeAsset(req.asset);

    const evmNetwork = CAIP_TO_CDP_EVM_BALANCE_NETWORK[req.network];
    const evmRpcUrl = rpcUrls[req.network]?.rpcUrl;
    const svmNetwork = CAIP_TO_CDP_SVM_BALANCE_NETWORK[req.network];

    let available: bigint;
    let address: string;

    const assetLabel = svmNetwork ? req.asset : normalizedAsset;

    try {
      if (evmNetwork) {
        address = evmAddress;
        available = await getEvmBalance(cdpClient, {
          address: evmAddress,
          network: evmNetwork,
          asset: normalizedAsset,
        });
      } else if (evmRpcUrl) {
        if (!normalizedAsset.startsWith("0x")) return undefined;
        address = evmAddress;
        available = await readOnChainErc20Balance(
          evmRpcUrl,
          normalizedAsset as `0x${string}`,
          evmAddress,
        );
      } else if (svmNetwork) {
        address = svmAddress;
        available = await getSvmBalance(cdpClient, {
          address: svmAddress,
          network: svmNetwork,
          asset: req.asset,
        });
      } else {
        return undefined;
      }
    } catch (error) {
      warn(
        `Pre-flight balance check failed for ${req.network} (${assetLabel}); ` +
          "proceeding without check.",
        error,
      );
      return undefined;
    }

    if (available >= required) return undefined;

    throw new InsufficientFundsError({
      required,
      available,
      asset: req.asset,
      network: req.network,
      address,
    });
  };
}
